### DISCUSSION TITLE
Non-root generic middlewares will not trigger if it does not explicitly contain routes

### DISCUSSION THREAD

**Discussion by BobbyWibowo** [kartikk221/hyper-express#110] [Q&A] [answered]
Suppose the following codes:

```js
// const express = require('express')
const HyperExpress = require('hyper-express')

console.log('Starting\u2026')
// const server = express()
const server = new HyperExpress.Server({
  fast_abort: true
})

server.use((req, res, next) => {
  console.log('root middleware')
  next()
})

server.use('/swagger', (req, res, next) => {
  console.log('/swagger middleware')
  next()
})

server.set_not_found_handler((req, res) => {
  console.log('not_found_handler')
  res.status(404).end()
})

server.set_error_handler((req, res) => {
  console.log('error_handler')
  res.status(500).end()
})

;(async () => {
  const port = process.env.PORT || 3001
  await server.listen(port)
  console.log(`Started on port ${port}`)
})()
```

When you do  `GET /swagger/anything`, hyper-express will only trigger the root middleware

This means you cannot have a generic middleware on non-root routes designed to selectively handle requests within non-root routes
For example, `serve-static`-alike middleware

You could add:
```js
server.any('/swagger/*', (req, res) => {})
```
Which will then allow the swagger middleware to trigger properly,  but that's more of a workaround than anything

You will still have to programmatically call your not_found_handler directly if your middleware can't handle the request due to not found file and such (i.e. can't just do `next()`), since that `any()` handler will simply continue to handle the request instead

---

On another note, you indeed can just do the following in ExpressJS
```js
server.use('/non-root-path', express.static(path))
```
Because following the code example above, swagger middleware will simply trigger properly when doing `GET /swagger/anything`

**Comment by kartikk221** [comment, marked as answer]
Great find, this makes total sense as Routes are ones that contain and execute middlewares thus If no actual routes exist for a specific range of pattern, then those middlewares will never run. I'll add a fix for this to improve this behavior in the next update.

**Reply by kartikk221** [reply]
Hi @BobbyWibowo could you see If the new v6.4.6 update resolves your issue with middlewares not running on your above requests?

**Reply by BobbyWibowo** [reply]
I can confirm that for this issue specifically, v6.4.6 pretty much solves it 👍 

---

I think it's just missing the ability for the dynamic middleware to be aware of the route context of the incoming requests, automatically

As in, when you initiate your dynamic middleware with the following:
```js
safe.use('/swagger', staticMiddleware(path))
```
The middleware cannot initially tell that all the requests that it will handle will actually have their `Request.path` always be prefixed with `/swagger`, since naturally only requests to that path will be passed to the middleware

Thus meaning, when initiating the dynamic middleware, you kinda have to manually inform the middleware of the context, such as adding support for the following in your middleware:
```js
safe.use('/swagger', staticMiddleware(path, { root: '/swagger' })
```
You need that context since naturally you can't just try to match files in the specified `path` with `Request.path`

For example, if `path` is `/path/to/swagger-dir`, then assume request to `http://localhost/swagger/index.html`
Its `Request.path` will be `/swagger/index.html` which can't be matched to `/path/to/swagger-dir/index.html` before first getting rid of the `/swagger` part

---

From what I can tell, in ExpressJS 4.18.1, if you initiate the middlewares the following way:
```js
server.use((req, res, next) => {
  console.log('root middleware:', req.path)
  next()
})

server.use('/swagger', (req, res, next) => {
  console.log('/swagger middleware:', req.path)
  next()
})
```
When initiating request to `http://localhost/swagger/index.html`, request passed to the `/swagger` middleware will actually have its `Request.path` property automatically be `/index.html` only
Meaning ExpressJS actually takes care of automatically removing it, as it arrives to the relevant middleware

![image](https://i.fiery.me/0d3m.png)

**Reply by kartikk221** [reply]
So essentially Express will modify the `Request.path` to reflect it's local path to the middleware's globalp path instead of the 
server root global request path?

**Reply by BobbyWibowo** [reply]
Apparently so, yeah

In the context of serve-static-alike middleware, that behavior makes the middleware just work, regardless of the actual path it's being initiated on, so I guess that makes sense

I cannot say if matching that behavior is fine for every other scenarios however
At least for my own use case it'll only bring benefits

**Reply by kartikk221** [reply]
Changing the behavior of `Request.path` would certainly be a breaking change for any existing implementations people may have on HyperExpress. So I'll add it to the roadmap to be implemented in the next major switch.

**Reply by BobbyWibowo** [reply]
Sounds good to me
At least 6.4.6 no longer require the `server.any()`-thingy workaround, and the cons that come with doing so, which is already pretty good imho

**Reply by kartikk221** [reply]
> Sounds good to me At least 6.4.6 no longer require the `server.any()`-thingy workaround, and the cons that come with doing so, which is already pretty good imho

Yeah, I was too focused on performance and would essentially match all registered middlewares to all registered routes based on path. Never occured to me that wildcard routes can receive requests to any set of paths hence per-request path pattern matching would have to be done to cover 100% of requests with middlewares.
